Modules/_asynciomodule.c

```diff
@@ -1012,7 +1012,13 @@ _asyncio_Future_remove_done_callback_impl(FutureObj *self, PyTypeObject *cls,
     ENSURE_FUTURE_ALIVE(state, self)
 
     if (self->fut_callback0 != NULL) {
-        int cmp = PyObject_RichCompareBool(self->fut_callback0, fn, Py_EQ);
+        // Beware: An evil PyObject_RichCompareBool could change fut_callback0
+        // (see https://github.com/python/cpython/issues/125789 for details)
+        // In addition, the reference to self->fut_callback0 may be cleared,
+        // so we need to temporarily hold it explicitly.
```

Comment:
This comment feels a bit confusing. I had a hard time figuring where in that issue to look for the example involving callback0. I also had a hard time understanding that returning `NotImplemented` will cause a second rich comparison to be invoked (by the first one) after the first argument to the first one has been DECREF'ed. (Thanks to the long comments below I now understand this -- good catch all!). Maybe deep-link to a specific comment in the issue? Or maybe just explain that if `fn`'s `__eq__` is evil enough, it can cause the first arg to PyObject_RichCompareBool  be freed before a recursive PyObject_RichCompareBool  calls is made with that same arg.

But now I am thinking that this is really a general weakness in PyObject_RichCompareBool that should be fixed there! (Maybe in do_richcompare.) Thoughts?

Reply:
> Or maybe just explain that if fn's __eq__ is evil enough, it can cause the first arg to PyObject_RichCompareBool be freed before a recursive PyObject_RichCompareBool calls is made with that same arg.

That's a good way to phrase it; I also didn't really know how to say it properly. Alternatively, I think a link to the comment https://github.com/python/cpython/pull/125967#discussion_r1816593340 would be fine as well though it this will make the line exceed 80 characters so it might look a bit ugly. What do you think?

> But now I am thinking that this is really a general weakness in PyObject_RichCompareBool that should be fixed there! (Maybe in do_richcompare.) Thoughts?

Unfortunately, I'm not sure we can always make it work. It would be great if `PyObject_RichCompareBool` was smart enough to figure it out but there are cases when it would not be straightforward (for instance: https://github.com/python/cpython/issues/119004#issuecomment-2108428409). Note that the evil `__eq__` works in this specific situation because we can mutate the operands being compared to trigger the UAF but that's not always the case I think =/

Reply:
By the way, there are at least two ways of fixing a UAF like this: either you incref/decref eagerly or you transfer ownership. The latter is only done if you'll anyway clear the ref after the call and is cheaper than the former, but in general, we need to do the latter. Since `__eq__` is a common operation, we try to be as efficient possible.